src/time.rs

```diff
@@ -0,0 +1,281 @@
+        write!(f, "{}.{}s", self.secs(), self.millis())
+    }
+}
+
+impl ops::Add<Duration> for Instant {
+    type Output = Instant;
+
+    fn add(self, rhs: Duration) -> Instant {
+        Instant::from_millis(self.millis + rhs.total_millis() as i64)
+    }
+}
+
+impl ops::Sub<Duration> for Instant {
+    type Output = Instant;
+
+    fn sub(self, rhs: Duration) -> Instant {
+        Instant::from_millis(self.millis - rhs.total_millis() as i64)
+    }
+}
+
+/// A relative amount of time.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Duration {
+    millis: u64,
+}
+
+impl Duration {
+    /// Create a new `Duration` from a number of milliseconds.
+    pub fn from_millis(millis: u64) -> Duration {
+        Duration { millis }
+    }
+
+    /// The fractional number of milliseconds in this `Duration`.
+    pub fn millis(&self) -> u64 {
+        self.millis % UNSIGNED_MILLIS_PER_SEC
+    }
+
+    /// The number of whole seconds in this `Duration`.
+    pub fn secs(&self) -> u64 {
+        self.millis / UNSIGNED_MILLIS_PER_SEC
+    }
+
+    /// The total number of milliseconds in this `Duration`.
+    pub fn total_millis(&self) -> u64 {
+        self.millis
+    }
+}
+
+impl fmt::Display for Duration {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}.{}s", self.secs(), self.millis())
+    }
+}
+
+impl ops::Add<Duration> for Duration {
+    type Output = Duration;
+
+    fn add(self, rhs: Duration) -> Duration {
+        Duration::from_millis(
+            self.millis.checked_add(rhs.total_millis()).expect("overflow when adding durations"))
+    }
+}
+
+impl ops::AddAssign<Duration> for Duration {
+    fn add_assign(&mut self, rhs: Duration) {
+        self.millis = self.millis.checked_add(rhs.total_millis()).expect("overflow when adding durations");
+    }
+}
+
+impl ops::Sub<Duration> for Duration {
+    type Output = Duration;
+
+    fn sub(self, rhs: Duration) -> Duration {
+        Duration::from_millis(
+            self.millis.checked_sub(rhs.total_millis()).expect("overflow when subtracting durations"))
+    }
+}
+
+impl ops::SubAssign<Duration> for Duration {
+    fn sub_assign(&mut self, rhs: Duration) {
+        self.millis = self.millis.checked_sub(
+            rhs.total_millis()).expect("overflow when subtracting durations");
+    }
+}
+
+impl ops::Mul<u32> for Duration {
+    type Output = Duration;
+
+    fn mul(self, rhs: u32) -> Duration {
+        Duration::from_millis(
+            self.millis.checked_mul(rhs as u64).expect("overflow when multiplying by scalar"))
+    }
+}
+
+impl ops::MulAssign<u32> for Duration {
+    fn mul_assign(&mut self, rhs: u32) {
+        self.millis = self.millis.checked_mul(rhs as u64).expect("overflow when multiplying by scalar");
+    }
+}
+
+impl ops::Div<u32> for Duration {
+    type Output = Duration;
+
+    fn div(self, rhs: u32) -> Duration {
+        Duration::from_millis(self.millis / rhs as u64)
```

Comment:
Conversely, what libstd does and what you forgot to do here is to check for division by zero.